### PR TITLE
Service factories

### LINKED ISSUE #43: Service Factories
This would be a nice feature addition:

```php
$config = [
    'db' => [
        'host'     => 'localhost',
        'database' => 'example_db',
        'username' => 'example_user',
        'password' => 'example_password',
    ],

    'di' => [
        'services' => [
            'database' => [
                'factory'   => MySQLPDOFactory::class,
                'singleton' => true,
                'arguments' => [
                    'config.db.host',
                    'config.db.database',
                    'config.db.username',
                    'config.db.password',
                ],
            ],
        ],
    ],
];
```

```php
<?php

class MySQLPDOFactory
{
    public function __invoke($host, $database, $username, $password)
    {
        $dsn = "mysql:host=$host;dbname=$database";
        $pdo = new PDO($dsn, $username, $password);
        $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

        return $pdo;
    }
}
```

### PR DESCRIPTION
**Pull request by tomphp**
Resolves #43

### CODE REVIEW

**Comment by tomphp** [comment]
@localheinz there's some questionable naming in this PR - feel free to be picky ;-)

File: README.md [hunk 1]
```diff
@@ -184,6 +184,59 @@ Configurator::apply()->configFromArray($config)->to($container);
 $logger = $container->get('logger'));
 ```
 
+#### Service Factories
+
+If you require some addition additional logic when creating a service, you can
+define a Service Factory. A service factory is simply and invokable class which
```
Comment (localheinz) [inline README.md]: @tomphp 

Should this be

```diff
-define a Service Factory. A service factory is simply and invokable class which
+define a Service Factory. A service factory is simply an invokable class which
```

instead?

File: README.md [hunk 2]
```diff
@@ -184,6 +184,59 @@ Configurator::apply()->configFromArray($config)->to($container);
...
+##### Example Config
+```php
+$appConfig = [
+    'db' => [
+        'host'     => 'localhost',
+        'database' => 'example_db',
+        'username' => 'example_user',
+        'password' => 'example_password',
+    ],
+
```
Comment (localheinz) [inline README.md]: @tomphp 

Should we remove this extra empty line, or is it here on purpose?

File: src/League/ServiceServiceProvider.php [hunk 3]
```diff
@@ -42,6 +42,16 @@ public function register()
      */
     private function registerService(ServiceDefinition $definition)
     {
+        if ($definition->isFactory()) {
+            $service = $this->getContainer()->add(
```
Comment (localheinz) [inline src/League/ServiceServiceProvider.php]: @tomphp 

After assigning `$service` a value here, it is no longer used, should we remove it?

File: src/ServiceDefinition.php [hunk 4]
```diff
@@ -23,6 +24,11 @@
     private $singleton;
 
     /**
+     * @var bool
+     */
+    private $factory;
```
Comment (localheinz) [inline src/ServiceDefinition.php]: @tomphp 

Consistent with what we got, but should we consider prepending `is` to these boolean fields?

Wasn't quite sure at first whether this was supposed to be the name of the factory.

File: tests/acceptance/SupportsServiceConfig.php [hunk 5]
```diff
@@ -287,4 +288,32 @@ public function testItUsesTheStringIffSetterMethodArgumentsAreClassNames()
 
         assertSame(ExampleClass::class, $instance->getValue());
     }
+
+    public function testIsCreatesAServiceThroughAFactoryClass()
+    {
+        $config = [
+            'class_name' => ExampleClassWithArgs::class,
+
```
Comment (localheinz) [inline tests/acceptance/SupportsServiceConfig.php]: @tomphp 

Should we remove this extra empty line, or is it here on purpose?

File: README.md (hunk 2 above)
Comment (tomphp) [inline README.md]: I guess it was intentional but happy to remove it.

**Comment by tomphp** [comment]
@localheinz I've addressed the issues but it seems a forced push doesn't work with the new GitHub code review stuff - will bare that in mind in the future.

**Comment by localheinz** [comment]
@tomphp 

Yes, it's a bit weird that the comments are still there, noticed it, too!

**Pull request merged**